### PR TITLE
feat: warn system

### PR DESCRIPTION
**Pull request by benw202**
## Description

Implements a warn system

- [x] Warn
- [x] List warnings
- [x] Delete warnings

## Test Results

N/A yet

## Discord Username

BenW#8484

### CODE REVIEW

File: src/commands/moderation/warn/deletewarn.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    color: 'RED',
+});
+
+const noWarning = makeEmbed({
+    title: 'Warn - No Warning',
+    description: 'Could not find warning. Please check the `Warn ID`',
+    color: 'RED',
+});
+
+export const deletewarn: CommandDefinition = {
```
Comment (NathanInnes) [inline src/commands/moderation/warn/deletewarn.ts:35]: Camelcase filename and const

**Review by NathanInnes — CHANGES_REQUESTED** [review]

File: src/commands/moderation/warn/warnings.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import moment from 'moment';
+import { CommandDefinition } from '../../../lib/command';
+import { Roles, CommandCategory } from '../../../constants';
+import { makeEmbed } from '../../../lib/embed';
+import { getConn } from '../../../lib/db';
+
+const permittedRoles = [
+    Roles.ADMIN_TEAM,
+    Roles.MODERATION_TEAM,
+];
```
[the remark this answers is not in the document]
Comment (NathanInnes) [reply src/commands/moderation/warn/warnings.ts:11]: Maybe for further clarification, you may want to also rename the file to "listWarnings".

File: src/commands/moderation/warn/deletewarn.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import { CommandDefinition } from '../../../lib/command';
+import { Roles, CommandCategory } from '../../../constants';
+import { makeEmbed } from '../../../lib/embed';
+import { getConn } from '../../../lib/db';
+
+const permittedRoles = [
+    Roles.ADMIN_TEAM,
+    Roles.MODERATION_TEAM,
+];
```
[the remark this answers is not in the document]
Comment (benw202) [reply src/commands/moderation/warn/deletewarn.ts:9]: Would this be beneficial? We use a different combination of roles for different commands, so would have to have all those combinations in constants

File: src/commands/moderation/warn/deletewarn.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    color: 'RED',
+});
+
+const deleteEmbed = makeEmbed({
+    title: 'Warn - Removed',
+    description: 'Warning has been removed',
+    color: 'GREEN',
+});
+
+const removeFailed = makeEmbed({
```
[the remark this answers is not in the document]
Comment (benw202) [reply src/commands/moderation/warn/deletewarn.ts:23]: No reason, draft PR. I haven't been through it yet. This is just it's rough and ready implementation

File: src/commands/moderation/warn/deletewarn.ts (hunk 1 above)
Comment (benw202) [reply src/commands/moderation/warn/deletewarn.ts:35]: We have a mixture of camelcase and non - will have to refactor the rest

File: src/commands/moderation/warn/deletewarn.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (benw202) [reply src/commands/moderation/warn/deletewarn.ts]: Haha okay

File: src/lib/db.ts [hunk 5]
```diff
@@ -12,6 +13,7 @@ export async function connect(url: string, callback = Logger.error) {
 
         // Register schemas
         connection.model('Birthday', birthdaySchema);
+        connection.model('Warn', warnSchema)
```
Comment (NathanInnes) [inline src/lib/db.ts:16]:
```suggestion
        connection.model('Warn', warnSchema);
```
Missing semi-colon.

**Review by NathanInnes — CHANGES_REQUESTED** [review]

File: src/commands/moderation/warn/warn.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+];
+
+const warnEmbed = (user: User) => makeEmbed({
+    title: `${user.tag} was warned successfully`,
+    color: 'GREEN',
+});
+
+const modLogEmbed = (formattedDate, moderator: User, user: User, reason: string) => makeEmbed({
+    author: {
+        name: `[warned]  ${user.tag}`,
```
Comment (NathanInnes) [inline src/commands/moderation/warn/warn.ts:20]:
```suggestion
        name: `[WARNED]  ${user.tag}`,
```
Better appearance on discord.

**Review by NathanInnes — COMMENTED** [review]
A few things I have found with testing.

- There is no ability to warn a user by mention.
- There is no log for removing a warn.
- A few more aliases for deletewarn and listwarnings would be nice.

**Comment by NathanInnes** [comment]
A further recommendation is to take a look at this PR #266 and possibly look at implementing the Schema this way?

**Review by NathanInnes — APPROVED** [review]

**Pull request merged**